ssc test speed up

sjanzou opened the pull request:
Low hanging fruit in the Google Doc completed.

The following results from running locally on Ubuntu machine with debug build for coverage results
Coverage reduced by 43 lines 63,733 to 63,690 (same percentage 54.7%)
Tests reduced from 776 to 690
Time reduced from 4,555.25s (75.92m) to 2,239.21s (37.32m)
Gory details at https://github.com/NREL/SAM-documentation/blob/main/Unit%20Testing/ssc%20test%20speedup/SSC%20Test%20speedup.docx

sjanzou committed:
eliminate PVSmoothing_lib_battery_dispatch.FuelCell_PV_Phoenix_all_on and check coverage

sjanzou committed:
Change pvsamv1 integration test pvsamv1_pv_defaults analysis_period from 25 to 2

sjanzou committed:
Update 6 failing tests when changing from 25 to 2 years for pvsamv1 defaults

sjanzou committed:
Additional changes of 25 years to 2 years in all analysis_period assignments

sjanzou committed:
All assigned analysis periods changed from 25 to 2 with updated tests

sjanzou committed:
Update NofinancialModuleSystemDesign and check coverage

sjanzou committed:
Test etes, ptes is_dispatch=1 only for reduction in lines covered

sjanzou committed:
Update ptes and etes tests to run all is_dispatch and update pvsamv1 integration tests

sjanzou committed:
coverage test for cashloan and singleowner

sjanzou committed:
reduce number of cash loan tests to check coverage

sjanzou committed:
another cashloan coverage test

sjanzou committed:
pvbatt/res/com and fuelcell/com and discounted payback cashloan tests

sjanzou committed:
parse down single owner tests

sjanzou committed:
benchmarking partnership flip tests

sjanzou committed:
reduce from 16 to 2 for all equity partnership flip tests

sjanzou committed:
add two tests back for additional coverage

sjanzou committed:
add wind back for different incentives

sjanzou committed:
testing biopower and pv battery for o and m variation

sjanzou committed:
add custom generation battery

sjanzou committed:
add mspt

sjanzou committed:
increasing coverage

sjanzou committed:
Reduce 32 tests to 12 with same coverage

for all partnership flip tests

sjanzou committed:
Update tests for cashloan and single owner for additional coverage

sjanzou committed:
Try to recover 9 lines of coverage

sjanzou committed:
Add PV back for single owner and cashloan

sjanzou committed:
full cash loan and single owner coverage with 16 instead of 43 tests

sjanzou committed:
final single owner and cash loan tests updates

tests reduced from 43 to 15 with same coverage and time reduced from 8573ms to 3940ms

sjanzou committed:
Baseline of remaining financial tests

sjanzou committed:
Merge remote-tracking branch 'origin/speedy_gonzales' into speedy_gonzales

sjanzou committed:
Update host developer tests for speedy_gonzales

sjanzou committed:
Third party ownership tests updates for speedy_gonzales

sjanzou committed:
sale leaseback tests updates for speedy gonzales

sjanzou committed:
lcoe fcr test updates for speedy Gonzales project

sjanzou committed:
Merchant Plant updates for speedy Gonzales project

Same coverage 8655 lines for TPO, HD, SL, MP, FCR compute module tests with number of tests reduced from 70 to 33 and time reduction from 24,182ms to 11,790ms

sjanzou committed:
Re-enable all tests and compare with GitHub actions

janinefreeman commented:
@tyneises @cpaulgilman @brtietz @dguittet @mjprilliman We discussed having all of us review on an extended timeline. Please review the **text** changes for technologies relevant to you in the test changes, and approve the PR when you've done so, but we do not all need to actually run the tests locally. Please aim for COB 5/30 for review. @cpaulgilman could you please be the one to run the actual test suite locally on your machine and make sure they pass?

dguittet reviewed (approved):
Awesome such simple changes made a big difference. Please remove all the comments though

brtietz commented on test/ssc_test/cmod_pvsamv1_test.cpp:
```diff
@@ -666,7 +670,7 @@ TEST_F(CMPvsamv1PowerIntegration_cmod_pvsamv1, NoFinancialModelSystemDesign)
     };
 
-    for (size_t i = 0; i != annual_energy_expected.size(); i++)
+    for (size_t i = 57; i < annual_energy_expected.size(); i++)
```
I'm looking at coveralls and trying to find the coverage decrease, and given that the mismatch losses are no longer covered I think this a likely candidate:

![image](https://github.com/user-attachments/assets/f1cfd6e1-60f8-4aa7-b2e9-e42c6cc10bc0)

Mismatch losses appear to be case i = 13. Can you:

(a) Edit the arrays above to only include the cases we want to keep? 
(b) Ensure mismatch losses are included in those cases?

sjanzou replied:
Great catch - running i=13 gives the same coverage as running all 64 configurations

base run (no changes from the i=57 to annual_energy_expected.size()
![image](https://github.com/user-attachments/assets/0dcb9811-9d77-4725-9dc9-017171e47f19)

i=13 results - coverage increased and time reduced
![image](https://github.com/user-attachments/assets/fc30f174-5a7b-4065-b5fb-4c8842bd1f14)

enabling all tests does not increase coverage and a 2.5 times increase in elapsed time
![image](https://github.com/user-attachments/assets/2e23828f-f8f2-4a19-a2b3-116d148125c4)

Updated in latest commit and commented code to be removed after one additional test

brtietz reviewed (commented):
Looks good overall, this is a very impressive speedup! I requested one change where I think the coverage loss is significant.

janinefreeman commented on test/input_cases/geothermal_common_data.h:
```diff
@@ -146,6 +146,7 @@ static void geothermal_singleowner_default(ssc_data_t &data)
 	ssc_data_set_number(data, "grid_interconnection_limit_kwac", 100000);
 	
 	ssc_data_set_number(data, "analysis_period", 25);
+
```
Was there a reason this one is still set to 25? Maybe doesn't actually really affect test time if geothermal doesn't run in lifetime mode, but worth a comment if that's the case

sjanzou replied:
@janinefreeman, thanks for the feedback - originally left at 25 because the speed impact was low
![image](https://github.com/user-attachments/assets/c4ec9077-3263-4810-867b-0d148d8e872b)

Updated to 2 years for consistency. However, this caused a failure in the cmod_grid 
![image](https://github.com/user-attachments/assets/6bcde3d9-c74e-402c-8b49-2e892096f327)

The issue was when the "analysis_period" is changed from 25 to 2, then the "geothermal_analysis_period" has to also be changed from 25 to 2
![image](https://github.com/user-attachments/assets/8f71f50c-6a0a-4d9c-b29e-3e57d628c25f)

Then everything runs
![image](https://github.com/user-attachments/assets/7264c6c8-a694-4ff7-b2ff-c16500802d55)

janinefreeman reviewed (approved):
Thanks Steve! Only one comment from me.

mjprilliman commented on test/ssc_test/cmod_pvsamv1_test.cpp:
```diff
@@ -145,11 +145,13 @@ TEST_F(CMPvsamv1PowerIntegration_cmod_pvsamv1, DefaultResidentialModel)
         ssc_data_get_number(data, "lcoe_nom", &lcoe_nom);
-        EXPECT_NEAR(lcoe_nom, 7.03, m_error_tolerance_lo) << "Levelized COE (nominal)";
+//        EXPECT_NEAR(lcoe_nom, 7.03, m_error_tolerance_lo) << "Levelized COE (nominal)";
+        EXPECT_NEAR(lcoe_nom, -26.45, m_error_tolerance_lo) << "Levelized COE (nominal)"; // Speedy Gonzales 25yr to 2yr
```
Negative LCOE? Might want to look at defaults for 2 year case, just to keep a meaningful value to check in the future. For 2 year runs, I set itc_fed_percent and itc_sta_percent to = [0] in pvsamv1_common_data.cpp and got a more reasonable lcoe (but a negative npv).

sjanzou replied:
lcoe changed from -26.45 to 12.05 with the itc_fed_percent = 0 and itc_sta_percent = 0 in latest commit

mjprilliman commented on test/ssc_test/cmod_pvsamv1_test.cpp:
```diff
@@ -145,11 +145,13 @@ TEST_F(CMPvsamv1PowerIntegration_cmod_pvsamv1, DefaultResidentialModel)
         ssc_data_get_number(data, "lcoe_real", &lcoe_real);
-        EXPECT_NEAR(lcoe_real, 5.65, m_error_tolerance_lo) << "Levelized COE (real)";
+//        EXPECT_NEAR(lcoe_real, 5.65, m_error_tolerance_lo) << "Levelized COE (real)";
+        EXPECT_NEAR(lcoe_real, -25.50, m_error_tolerance_lo) << "Levelized COE (real)"; // Speedy Gonzales 25yr to 2yr
```
Same comment as above

sjanzou replied:
lcoe changed from -25.5 to 11.61 with the itc_fed_percent = 0 and itc_sta_percent = 0 in latest commit

npv changed from 5685.3 to -358.2

mjprilliman commented on test/ssc_test/cmod_singleowner_test.cpp:
```diff
@@ -102,7 +102,7 @@ TEST_F(CmodSingleOwnerTest, ETES) {
 }
-
+/*
 TEST_F(CmodSingleOwnerTest, PV) {
```
As PV-SingleOwner is likely one of the most modeled technologies, I might consider keeping this one and commenting out others. Although I see PV Battery is still being run, so there may be enough coverage already. Same thought for PVWatts-SingleOwner

sjanzou replied:
The ETES test was kept because it uses an additional o and m value that the others single owner tests do not. Same with the battery and fuel cell. The other tests were selected based on inputs to single owner compute module being changed. The PV and PVwatts single owner tests were added back (no coverage change and only an additional 40ms added)

mjprilliman reviewed (approved):
Just a few nitpicks, thanks for the great work on this!

sjanzou committed:
Merge branch 'patch' into speedy_gonzales

sjanzou committed:
Address feedback from @mjprilliman

sjanzou committed:
Address all comments from @brtietz and increase coverage

Remaining items
1.	Remove all but i=13 tests from pvsamv1 integration tests
2.	Remove commented code
3.	Merge pull request

brtietz reviewed (approved):
Thanks for updating the pvsamv1 tests!

sjanzou committed:
Merge branch 'patch' into speedy_gonzales

sjanzou committed:
Remove all commented out tests and address all comments on #1327

sjanzou commented:
Latest coverage results from local Ubuntu VM
![image](https://github.com/user-attachments/assets/f8792d1f-6881-435f-8c4c-f38b3cdaa95a)
